### ISSUE TITLE
[Login] Promise-based methods don't return subscription list

### ISSUE THREAD

**Issue by lostintangent** [Azure/azure-sdk-for-node#2115]
I might be missing something, but I noticed that when using the `login*` methods, the new `Promise` support doesn't expose the list of subscriptions, like the callbacks do. Is there a way to use `Promises` and get both the credentials and subscription list after a successful login?

**Comment by amarzavery** [comment]
Subscriptions was just an extra thing that we were returning. We don't expect customers to use it. The usecase is that the customer knows which subscription to use. 

Do you see any use for subscriptions?

**Comment by lostintangent** [comment]
In my usage, I'm allowing the user to login to a tool interactively (via the device code workflow), and then I auto-select the subscription to use (if they only had one), or prompt them to select the subscription they'd like to use (if they had multiple). I then generate an SP to use for subsequent runs of the tool, as opposed to requiring the interactive workflow again.

This has the benefit of not requiring the user to pre-configure anything, and the tool can be more "guided" about selecting the subscription.

**Comment by lostintangent** [comment]
It's not a big deal to use the callback solution, I just thought I'd file this issue to have the conversation.

**Comment by balajikris** [comment]
The current design lets us write code like so, which is neat.

```ts
var credentials = await login();
```

If there is a use case that requires the `login` call to return both `credentials` and `subscriptions`, we could consider designing an API for it.

**Comment by lostintangent** [comment]
@balajikris Didn't I just mentioned a use case? :) If the object returned by the `login` methods returned both the `credentials` and `subscriptions`, you could still write the following code to pull out just the `credentials` if you wanted.

```javascript
const { credentials } = await login();
```

**Comment by balajikris** [comment]
Yes you did :) I was asserting that if there is a usecase, we should indeed consider a solution for it :) Falling back to callbacks in a promise based codebase points out to areas where we can improve, imo. 

I'm sure there will be more places in the API surface like this. It requires a bit of design as we could return only a promise of a single type `T` ?
Using union types like `: Promise<T1 | T2>` wasn't playing well with `then` chaining etc.

**Comment by balajikris** [comment]
@lostintangent just curious, what'd be the return type in the signature of `login` which lets you deconstruct `creds` like in your example? Forgive my asking, I'm slightly new to TS and JS.

**Comment by amarzavery** [comment]
- Deconstructing tuples requires special knowledge as user's wouldnt be expecting that. That is why we provided an overloaded method `client.someOperationGroup.someMethodWithHttpOperationResponse(): Promise<HttpOperationResponse<T>`.

A simple solution would be to provide an overloaded method as above. The method signature would look like this:
```javascript
msRestAzure.loginWithServicePrincipalSecretWithAuthResponse(spn: string, domain: string, secret: string): Promise<AuthResponse>;
```
where `AuthResponse` looks like this:

```javascript
export interface AuthResponse {
  credentials: msRest.ServiceClientCredentials;
  subscriptions: Array<Models.Subscription>
}
```

I did think about the above approach. However, wasn't sure if someone would be using the subscriptions. Hence, canned it. If people like this solution then we can have this out in a day or two.

In fact, I was planning to remove the azure-arm-resource dependency in ms-rest-azure and move the subscription client as a local client in ms-rest-azure. This would also remove circular dependency.

What do you guys think about it?

Note:
> One more thing, that is tangential to this conversation. At some point in the future, when we move to generating isomorphic javascript, we shall move login methods from `ms-rest-azure` and put it in a separate package named `ms-rest-azurenodeauth` (node.js auth) that depends on `ms-rest` and `ms-rest-azurejsauth` (client side js auth) that depends on `ms-rest`.

**Comment by lostintangent** [comment]
That proposal sounds good to me!

**Comment by amarzavery** [comment]
Closing this issue as the PR has been merged and new version of [ms-rest-azure: 2.1.0](https://github.com/Azure/azure-sdk-for-node/blob/master/runtime/ms-rest-azure/Changelog.md#210-452017) has been published.
